src/compiler/checker.ts

```diff
@@ -6568,6 +6616,394 @@ module ts {
+            }
+
+            checkGrammarJsxElement(node);
+
+            if (!jsxElementType) {
+                error(node, Diagnostics.The_global_type_JSX_Element_must_exist_when_using_JSX);
+                return unknownType;
+            }
+
+            var elemType = checkJsxOpeningElement(node.openingElement);
+
+            for (var i = 0, n = node.children.length; i < n; i++) {
+                if (node.children[i].kind === SyntaxKind.JsxExpression) {
+                    checkJsxExpression(<JsxExpression>node.children[i]);
+                }
+                else if (node.children[i].kind === SyntaxKind.JsxElement) {
+                    checkJsxElement(<JsxElement>node.children[i]);
+                }
+                else {
+                    Debug.assert(node.children[i].kind === SyntaxKind.JsxText);
+                }
+            }
+
+            if (node.closingElement) {
+                checkSourceElement(node.closingElement);
+            }
+
+            return jsxElementType;
+        }
+
+        /// Returns true iff the JSX element name would be a valid identifier, ignoring restrictions about keywords not being identifiers
+        function isIdentifierLike(name: string) {
+            // - is the only character supported in JSX attribute names that isn't valid in JavaScript identifiers
+            return name.indexOf('-') < 0;
+        }
+
+        /// Returns true iff React would emit this tag name as a string rather than an identifier or qualified name
+        function isJsxIntrinsicIdentifier(tagName: Identifier|QualifiedName) {
+            if (tagName.kind === SyntaxKind.QualifiedName) {
+                return false;
+            } else {
+                let firstChar = (<Identifier>tagName).text.charAt(0);
+                return firstChar.toLowerCase() === firstChar;
+            }
+        }
+
+        function checkJsxAttribute(node: JsxAttribute, elementAttributesType: Type, nameTable: Map<boolean>) {
+            var correspondingPropType: Type = unknownType;
+
+            // Look up the corresponding property for this attribute
+            if (elementAttributesType === emptyObjectType && isIdentifierLike(node.name.text)) {
+                // If there is no 'props' property, you may not have non-"data-" attributes
+                error(node.parent, Diagnostics.JSX_element_class_does_not_support_attributes_because_it_does_not_have_a_0_property, getJsxElementPropertiesName());
+            }
+            else if (elementAttributesType && !(elementAttributesType.flags & TypeFlags.Any)) {
+                var correspondingPropSymbol = getPropertyOfType(elementAttributesType, node.name.text);
+                correspondingPropType = correspondingPropSymbol && getTypeOfSymbol(correspondingPropSymbol);
+                if (!correspondingPropType) {
+                    // If there's no corresponding property with this name, error
+                    if (isIdentifierLike(node.name.text)) {

```

Comment:
Why does isIdentifierLike matter? What if your props type has a property with a quoted name that has a minus in it? We should respect that.